File: crypto.py
import base64
import hashlib
import hmac
import os

from cryptography.hazmat.primitives import padding as sym_padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from secmem import SecureBuffer, wipe


class SymmetricKey:
    def __init__(self, raw: bytes | bytearray):
        if len(raw) != 64:
            raise ValueError(f"expected 64 bytes, got {len(raw)}")
        self._secure = SecureBuffer(raw)
        if isinstance(raw, bytearray):
            wipe(raw)

    @property
    def raw(self) -> bytearray:
        return self._secure.raw

    @property
    def enc_key(self) -> bytearray:
        return self._secure.raw[:32]

    @property
    def mac_key(self) -> bytearray:
        return self._secure.raw[32:]

    def to_b64(self) -> str:
        return base64.b64encode(bytes(self._secure)).decode()

    def close(self):
        self._secure.close()

    @classmethod
    def from_b64(cls, b64: str) -> "SymmetricKey":
        return cls(base64.b64decode(b64))

    @classmethod
    def generate(cls) -> "SymmetricKey":
        return cls(os.urandom(64))


def _decrypt_raw(enc_str: str, key: SymmetricKey) -> bytearray:
    t, rest = enc_str.split(".", 1)
    parts = rest.split("|")
    iv = base64.b64decode(parts[0])
    ct = base64.b64decode(parts[1])
    if len(parts) > 2:
        mac_got = base64.b64decode(parts[2])
        expected = hmac.new(bytes(key.mac_key), iv + ct, hashlib.sha256).digest()
        if not hmac.compare_digest(mac_got, expected):
            raise ValueError("MAC mismatch")
    dec = Cipher(algorithms.AES(bytes(key.enc_key)), modes.CBC(iv)).decryptor()
    padded = dec.update(ct) + dec.finalize()
    unpadder = sym_padding.PKCS7(128).unpadder()
    return bytearray(unpadder.update(padded) + unpadder.finalize())


def enc_string_encrypt(plaintext: str, key: SymmetricKey) -> str:
    iv = os.urandom(16)
    padder = sym_padding.PKCS7(128).padder()
    padded = padder.update(plaintext.encode()) + padder.finalize()
    ct = Cipher(algorithms.AES(bytes(key.enc_key)), modes.CBC(iv)).encryptor()
    encrypted = ct.update(padded) + ct.finalize()
    mac = hmac.new(bytes(key.mac_key), iv + encrypted, hashlib.sha256).digest()
    iv_b64 = base64.b64encode(iv).decode()
    ct_b64 = base64.b64encode(encrypted).decode()
    mac_b64 = base64.b64encode(mac).decode()
    return f"2.{iv_b64}|{ct_b64}|{mac_b64}"


def enc_string_decrypt(enc_str: str, key: SymmetricKey) -> str:
    raw = _decrypt_raw(enc_str, key)
    result = raw.decode()
    wipe(raw)
    return result


def enc_string_decrypt_bytes(enc_str: str, key: SymmetricKey) -> bytearray:
    return _decrypt_raw(enc_str, key)


def enc_string_to_dict(enc_str: str) -> dict:
    t, rest = enc_str.split(".", 1)
    parts = rest.split("|")
    d = {"encryptionType": int(t), "encryptedString": enc_str}
    if len(parts) >= 1:
        d["iv"] = parts[0]
    if len(parts) >= 2:
        d["data"] = parts[1]
    if len(parts) >= 3:
        d["mac"] = parts[2]
    return d


def dict_to_enc_string(d: dict) -> str:
    if s := d.get("encryptedString"):
        return s
    t = d.get("encryptionType", 2)
    return f"{t}.{d.get('iv', '')}|{d.get('data', '')}|{d.get('mac', '')}"

File: src/crypto.rs
use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, BlockEncryptMut, KeyIvInit};
use base64::{engine::general_purpose::STANDARD as B64, Engine};
use hmac::{Hmac, Mac};
use rand::RngCore;
use sha2::Sha256;
use zeroize::Zeroizing;

type Aes256CbcEnc = cbc::Encryptor<aes::Aes256>;
type Aes256CbcDec = cbc::Decryptor<aes::Aes256>;

pub struct SymmetricKey {
    raw: Zeroizing<Vec<u8>>,
}

impl SymmetricKey {
    pub fn new(data: Vec<u8>) -> Self {
        assert_eq!(data.len(), 64);
        Self { raw: Zeroizing::new(data) }
    }

    pub fn generate() -> Self {
        let mut buf = vec![0u8; 64];
        rand::thread_rng().fill_bytes(&mut buf);
        Self::new(buf)
    }

    pub fn raw(&self) -> &[u8] {
        &self.raw
    }

    pub fn enc_key(&self) -> &[u8] {
        &self.raw[..32]
    }

    pub fn mac_key(&self) -> &[u8] {
        &self.raw[32..]
    }
}

pub fn enc_string_encrypt(plaintext: &str, key: &SymmetricKey) -> String {
    let mut iv = [0u8; 16];
    rand::thread_rng().fill_bytes(&mut iv);

    let encrypted = Aes256CbcEnc::new(key.enc_key().into(), &iv.into())
        .encrypt_padded_vec_mut::<Pkcs7>(plaintext.as_bytes());

    let mut mac_input = Vec::with_capacity(16 + encrypted.len());
    mac_input.extend_from_slice(&iv);
    mac_input.extend_from_slice(&encrypted);

    let mut hmac = Hmac::<Sha256>::new_from_slice(key.mac_key()).unwrap();
    hmac.update(&mac_input);
    let mac = hmac.finalize().into_bytes();

    format!(
        "2.{}|{}|{}",
        B64.encode(iv),
        B64.encode(&encrypted),
        B64.encode(mac)
    )
}

pub fn enc_string_decrypt(enc_str: &str, key: &SymmetricKey) -> Result<String, &'static str> {
    let raw = enc_string_decrypt_bytes(enc_str, key)?;
    String::from_utf8(raw.to_vec()).map_err(|_| "invalid utf8")
}

pub fn enc_string_decrypt_bytes(enc_str: &str, key: &SymmetricKey) -> Result<Zeroizing<Vec<u8>>, &'static str> {
    let (_t, rest) = enc_str.split_once('.').ok_or("bad format")?;
    let parts: Vec<&str> = rest.split('|').collect();
    if parts.len() < 2 {
        return Err("bad format");
    }

    let iv = B64.decode(parts[0]).map_err(|_| "bad iv")?;
    let ct = B64.decode(parts[1]).map_err(|_| "bad ct")?;

    if parts.len() > 2 {
        let mac_got = B64.decode(parts[2]).map_err(|_| "bad mac")?;
        let mut hmac = Hmac::<Sha256>::new_from_slice(key.mac_key()).unwrap();
        let mut mac_input = Vec::with_capacity(iv.len() + ct.len());
        mac_input.extend_from_slice(&iv);
        mac_input.extend_from_slice(&ct);
        hmac.update(&mac_input);
        hmac.verify_slice(&mac_got).map_err(|_| "MAC mismatch")?;
    }

    let decrypted = Aes256CbcDec::new(key.enc_key().into(), iv.as_slice().into())
        .decrypt_padded_vec_mut::<Pkcs7>(&ct)
        .map_err(|_| "decrypt failed")?;

    Ok(Zeroizing::new(decrypted))
}

pub fn enc_string_to_json(enc_str: &str) -> serde_json::Value {
    let (t, rest) = enc_str.split_once('.').unwrap_or(("2", enc_str));
    let parts: Vec<&str> = rest.split('|').collect();
    let mut m = serde_json::Map::new();
    m.insert("encryptionType".into(), serde_json::json!(t.parse::<u32>().unwrap_or(2)));
    m.insert("encryptedString".into(), serde_json::json!(enc_str));
    if let Some(iv) = parts.first() {
        m.insert("iv".into(), serde_json::json!(iv));
    }
    if let Some(data) = parts.get(1) {
        m.insert("data".into(), serde_json::json!(data));
    }
    if let Some(mac) = parts.get(2) {
        m.insert("mac".into(), serde_json::json!(mac));
    }
    serde_json::Value::Object(m)
}

pub fn json_to_enc_string(v: &serde_json::Value) -> String {
    if let Some(s) = v.get("encryptedString").and_then(|s| s.as_str()) {
        return s.to_string();
    }
    let t = v.get("encryptionType").and_then(|t| t.as_u64()).unwrap_or(2);
    let iv = v.get("iv").and_then(|s| s.as_str()).unwrap_or("");
    let data = v.get("data").and_then(|s| s.as_str()).unwrap_or("");
    let mac = v.get("mac").and_then(|s| s.as_str()).unwrap_or("");
    format!("{t}.{iv}|{data}|{mac}")
}
